src/ucs/sys/topo.c

```diff
@@ -19,9 +53,125 @@ ucs_sys_bus_id_t ucs_sys_bus_id_unknown = { .domain   = 0xffff,
+{
+    ucs_status_t status;
+    ucs_sys_device_t sys_dev;
+
+    if (!ucs_topo_initialized) return;
+
+    kh_foreach_value(&ucs_topo_bus_to_sys_dev_hash, sys_dev, {
+        ucs_debug("sys_device_t %u not released", sys_dev);
+    })
+
+    kh_destroy_inplace(bus_to_sys_dev, &ucs_topo_bus_to_sys_dev_hash);
+
+    status = ucs_recursive_spinlock_destroy(&ucs_topo_bus_to_sys_dev_lock);
+    if (status != UCS_OK) {
+        ucs_warn("ucs_recursive_spinlock_destroy() failed: %s",
+                 ucs_status_string(status));
+    }
+
+    status = ucs_recursive_spinlock_destroy(&ucs_topo_sys_dev_to_bus_lock);
+    if (status != UCS_OK) {
+        ucs_warn("ucs_recursive_spinlock_destroy() failed: %s",
+                 ucs_status_string(status));
+    }
+
+    ucs_topo_initialized = 0;
+}
+
+static int ucs_topo_compare_bus_id(const ucs_sys_bus_id_t *bus_id1, 
+                                   const ucs_sys_bus_id_t *bus_id2)
+{
+    if ((bus_id1->domain == bus_id2->domain) &&
+        (bus_id1->bus == bus_id2->bus) &&
+        (bus_id1->slot == bus_id2->slot) &&
+        (bus_id1->function == bus_id2->function)) {
+        return 1;
+    }
+
+    return 0;
+}
+
 ucs_status_t ucs_topo_find_device_by_bus_id(const ucs_sys_bus_id_t *bus_id,
                                             ucs_sys_device_t *sys_dev)
 {
+    khiter_t hash_it;
+    ucs_topo_kh_put_t kh_put_status;
+
+    *sys_dev = UINT_MAX;
+
+    if (ucs_topo_compare_bus_id(bus_id, &ucs_sys_bus_id_unknown)) {
+        ucs_debug("found unknown device index %u for bus id %ld",
+                  *sys_dev, ucs_topo_get_bus_id_bit_repr(bus_id));
+        return UCS_OK;
+    }
+    
+    ucs_debug("find device index for bus id %ld", ucs_topo_get_bus_id_bit_repr(bus_id));
+
+    ucs_recursive_spin_lock(&ucs_topo_bus_to_sys_dev_lock);
+    hash_it = kh_put(bus_to_sys_dev/*name*/, 
+                     &ucs_topo_bus_to_sys_dev_hash /*pointer to hashmap*/,
+                     ucs_topo_get_bus_id_bit_repr(bus_id)/*key*/,
+                     &kh_put_status);
+
+    switch (kh_put_status) {
+        case UCS_KH_PUT_KEY_PRESENT:
+            *sys_dev = kh_value(&ucs_topo_bus_to_sys_dev_hash, hash_it);
+            ucs_debug("bus id %ld exists. sys_dev = %u",
+                      ucs_topo_get_bus_id_bit_repr(bus_id), *sys_dev);
+            break;
+        case UCS_KH_PUT_BUCKET_EMPTY:
+            *sys_dev = ucs_topo_sys_dev_to_bus_lookup.count;
+            kh_value(&ucs_topo_bus_to_sys_dev_hash, hash_it) = *sys_dev;
+            ucs_debug("bus id %ld doesn't exist. sys_dev = %u",
+                      ucs_topo_get_bus_id_bit_repr(bus_id), *sys_dev);
+
+            /* because of grabbing lock while holding another lock all other
+             * instances must make sure that the opposite pattern is not
+             * attempted to avoid deadlock */
+            ucs_recursive_spin_lock(&ucs_topo_sys_dev_to_bus_lock);
+            ucs_topo_sys_dev_to_bus_lookup.bus_arr[*sys_dev] = *bus_id; 
```

Comment:
ptr_array may still be a better match, because you would not need to bother about counters/indexes. Also we have `locked_ptr_array` guarded by spin lock, so you will not need to define another lock

Reply:
Will look into this then.

Reply:
@brminich Looked into use of `ptr_array_locked` utilities but it looks like this structure assumes that the pointer values inserted into the array and the index associated with are valid only until the lifetime of the pointer. What we want ideally is just `array_locked` because we want into insert values that last the whole lifetime of the process. For example, at some point we would insert a bus_id value concerning some pcie device and we don't want to remove this entry from the array nor want the entry to point to invalid memory even if the bus_id variable that gets populated is destroyed. If we still want to use `ptr_array_locked` for this, then we'd have to allocate some memory to which we copy bus_id values and then pass in the pointers to entries in this new array as part of insert operation. But this would need an external lock to access the new array. With the above two requirements, is it not better to keep current implementation?

Reply:
Adding changes I attempted based on your suggestion here: 
```diff
diff --git a/src/ucs/sys/topo.c b/src/ucs/sys/topo.c
index 2c6a87fe6..b69031d84 100644
--- a/src/ucs/sys/topo.c
+++ b/src/ucs/sys/topo.c
@@ -41,8 +41,8 @@ KHASH_INIT(bus_to_sys_dev /*name*/,
 static khash_t(bus_to_sys_dev) ucs_topo_bus_to_sys_dev_hash;
 static ucs_recursive_spinlock_t  ucs_topo_bus_to_sys_dev_lock;
-static ucs_topo_sys_dev_to_bus_arr_t ucs_topo_sys_dev_to_bus_lookup;
-static ucs_recursive_spinlock_t  ucs_topo_sys_dev_to_bus_lock;
+static ucs_ptr_array_locked_t ucs_topo_sys_dev_to_bus_array;
 static int ucs_topo_initialized = 0;
@@ -70,12 +70,24 @@ static ucs_bus_id_bit_rep_t ucs_topo_get_bus_id_bit_repr(const ucs_sys_bus_id_t
 void ucs_topo_init()
 {
+    ucs_status_t status;
+
     if (ucs_topo_initialized) return;
-    ucs_recursive_spinlock_init(&ucs_topo_sys_dev_to_bus_lock, 0);
-    ucs_recursive_spinlock_init(&ucs_topo_bus_to_sys_dev_lock, 0);
+    status = ucs_recursive_spinlock_init(&ucs_topo_bus_to_sys_dev_lock, 0);
+    if (status != UCS_OK) {
+        ucs_warn("ucs_recursive_spinlock_init failed");
+        return;
+    }
+
     kh_init_inplace(bus_to_sys_dev, &ucs_topo_bus_to_sys_dev_hash);
-    ucs_topo_sys_dev_to_bus_lookup.count = 0;
+
+    status = ucs_ptr_array_locked_init(&ucs_topo_sys_dev_to_bus_array,
+                                       "sys_dev_to_bus_array");
+    if (status != UCS_OK) {
+        ucs_warn("ucs_ptr_array_locked_init failed");
+        return;
+    }
     ucs_topo_initialized = 1;
 }
@@ -99,10 +111,9 @@ void ucs_topo_cleanup()
                  ucs_status_string(status));
     }
-    status = ucs_recursive_spinlock_destroy(&ucs_topo_sys_dev_to_bus_lock);
+    status = ucs_ptr_array_locked_cleanup(&ucs_topo_sys_dev_to_bus_array);
     if (status != UCS_OK) {
-        ucs_warn("ucs_recursive_spinlock_destroy() failed: %s",
-                 ucs_status_string(status));
+        ucs_warn("ucs_ptr_array_locked_cleanup failed");
     }
     ucs_topo_initialized = 0;
@@ -150,7 +161,7 @@ ucs_status_t ucs_topo_find_device_by_bus_id(const ucs_sys_bus_id_t *bus_id,
                       ucs_topo_get_bus_id_bit_repr(bus_id), *sys_dev);
             break;
         case UCS_KH_PUT_BUCKET_EMPTY:
-            *sys_dev = ucs_topo_sys_dev_to_bus_lookup.count;
+            *sys_dev = ucs_ptr_array_locked_get_size(&ucs_topo_sys_dev_to_bus_array);
             kh_value(&ucs_topo_bus_to_sys_dev_hash, hash_it) = *sys_dev;
             ucs_debug("bus id %ld doesn't exist. sys_dev = %u",
                       ucs_topo_get_bus_id_bit_repr(bus_id), *sys_dev);
@@ -158,10 +169,7 @@ ucs_status_t ucs_topo_find_device_by_bus_id(const ucs_sys_bus_id_t *bus_id,
             /* because of grabbing lock while holding another lock all other
              * instances must make sure that the opposite pattern is not
              * attempted to avoid deadlock */
-            ucs_recursive_spin_lock(&ucs_topo_sys_dev_to_bus_lock);
-            ucs_topo_sys_dev_to_bus_lookup.bus_arr[*sys_dev] = *bus_id;
-            ucs_topo_sys_dev_to_bus_lookup.count++;
-            ucs_recursive_spin_unlock(&ucs_topo_sys_dev_to_bus_lock);
+            ucs_ptr_array_locked_insert(&ucs_topo_sys_dev_to_bus_array, bus_id);
             break;
         case UCS_KH_PUT_BUCKET_CLEAR:
             /* not handled */
@@ -180,6 +188,10 @@ ucs_status_t ucs_topo_get_distance(ucs_sys_device_t device1,
                                    ucs_sys_device_t device2,
                                    ucs_sys_dev_distance_t *distance)
 {
+    ucs_sys_bus_id_t *bus_id1, *bus_id2;
+    ucs_ptr_array_locked_lookup(&ucs_topo_sys_dev_to_bus_array, device1, &bus_id1);
+    ucs_ptr_array_locked_lookup(&ucs_topo_sys_dev_to_bus_array, device2, &bus_id2);
+    ...
     return UCS_OK;
 }
```

In short, the lookup call in `ucs_topo_get_distance` relies on the original memory that `bus_id` pointed to at the point of insertion is still valid which may not be the case. LMK if I'm being clear.

Reply:
ok, i see what you mean. 
i'd ask @yosefe to look thru, before going to other comments

Reply:
i think ptr_array is not relevant since we need to search by bus_id

Reply:
@brminich Shall I leave the current changes unchanged then?